src/http/one/TeChunkedParser.cc

```diff
@@ -115,55 +116,60 @@ Http::One::TeChunkedParser::parseChunkSize(Http1::Tokenizer &tok)
-
-        // whole value part is optional. if no '=' expect next chunk-ext
-        if (ParseBws(tok) && tok.skip('=') && ParseBws(tok)) {
-
-            if (!skipKnown) {
-                if (ext.cmp("use-original-body",17) == 0 && tok.int64(useOriginBody, 10)) {
-                    debugs(94, 3, "Found chunk extension " << ext << "=" << useOriginBody);
-                    buf_ = tok.remaining(); // parse checkpoint
-                    continue;
-                }
-            }
+    while (parseOneChunkExtension(tok, skipKnown)) {
+        buf_ = tok.remaining(); // got one extension; there may be more
+    }
 
-            debugs(94, 5, "skipping unknown chunk extension " << ext);
+    if (skipLineTerminatorIfAny(tok)) {
+        buf_ = tok.remaining();
+        parsingStage_ = theChunkSize ? Http1::HTTP_PARSE_CHUNK : Http1::HTTP_PARSE_MIME;
+        return true; // reached the end of extensions (if any)
+    }
 
-            // unknown might have a value token or quoted-string
-            if (tok.quotedStringOrToken(value) && !tok.atEnd()) {
-                buf_ = tok.remaining(); // parse checkpoint
-                continue;
-            }
+    static const CharacterSet NonLF = (CharacterSet::LF).complement().rename("non-LF");
+    if (tok.skipAll(NonLF) && tok.skip('\n'))
+        throw TexcHere("cannot parse chunk extension"); // <garbage> CR*LF
+    return false; // need more data
+}
+
+bool
+Http::One::TeChunkedParser::parseOneChunkExtension(Tokenizer &tok, const bool skipKnown)
+{
+    if (!ParseBws(tok)) // Bug 4492: IBM_HTTP_Server sends SP after chunk-size
+        return false;
+
+    if (!tok.skip(';'))
+        return false;
+
+    if (!ParseBws(tok)) // Bug 4492: ICAP servers send SP before chunk-ext-name
+        return false;
+
+    SBuf ext;
+    if (!tok.prefix(ext, CharacterSet::TCHAR)) // chunk-ext-name
+        return false;
 
-            // otherwise need more data OR corrupt syntax
-            break;
+    if (ParseBws(tok) && tok.skip('=') && ParseBws(tok)) {
+        if (!skipKnown) {
+            if (ext.cmp("use-original-body", 17) == 0 && tok.int64(useOriginBody, 10)) {
+                debugs(94, 3, "found " << ext << '=' << useOriginBody);
+                return true;
+            }
         }
 
-        if (!tok.atEnd())
-            buf_ = tok.remaining(); // parse checkpoint (unless there might be more token name)
-    }
+        debugs(94, 5, "skipping unknown chunk extension " << ext);
 
-    if (skipLineTerminator(tok)) {
-        buf_ = tok.remaining(); // checkpoint
-        // non-0 chunk means data, 0-size means optional Trailer follows
-        parsingStage_ = theChunkSize ? Http1::HTTP_PARSE_CHUNK : Http1::HTTP_PARSE_MIME;
-        return true;
+        SBuf ignoredValue;
+        return tokenOrQuotedString(tok, ignoredValue, false);
     }
 
-    return false;
+    // either parsed a valueless chunk-ext or need more data to check for optional value
+    return (!tok.atEnd());
```

Comment:
Please note that this parsing function may return false after modifying its tokenizer parameter. As you know, this is often a bad idea.